feat: Add F5 TransportServer source

visokoo opened the pull request:
<!--
    Please read https://github.com/kubernetes-sigs/external-dns#contributing before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

Should completely fix: #2739 by adding in the `TransportServer` CRD as well. This is basically an exact copy of this [PR](https://github.com/kubernetes-sigs/external-dns/pull/3162) from 2 yrs ago, but adds the `TransportServer` instead. We can use the same code because the `TransportServer` is essentially the same thing as the `VirtualServer`, but it accepts other ports other than 80 and 443.

**Checklist**

- [X] Unit tests updated

visokoo commented:
@mloiseleur just want to make sure it's on your radar. It's adding a new source.

mloiseleur commented:
@visokoo Thanks for this PR. Would you please complete [current documentation about F5 sources](https://kubernetes-sigs.github.io/external-dns/latest/docs/sources/f5-virtualserver/) with this new source, and show how it works from user point of view ?

mloiseleur commented:
/ok-to-test

visokoo commented:
Added docs @mloiseleur, thank you for reviewing!

mloiseleur commented on docs/sources/f5-transportserver.md:
```diff
@@ -0,0 +1,33 @@
+# F5 Networks TransportServer Source
+This tutorial describes how to configure ExternalDNS to use the F5 Networks TransportServer Source. It is meant to supplement the other provider-specific setup tutorials.
```
```suggestion
# F5 Networks TransportServer Source

This tutorial describes how to configure ExternalDNS to use the F5 Networks TransportServer Source. It is meant to supplement the other provider-specific setup tutorials.
```

mloiseleur commented on source/f5_transportserver.go:
```diff
@@ -0,0 +1,213 @@
+		return nil, err
+	}
+
+	// Sort endpoints
```
Would you please explain in this comment why the endpoints are sorted ?

visokoo replied:
Don't think it's actually needed so removed it.

mloiseleur commented on source/f5_transportserver.go:
```diff
@@ -0,0 +1,213 @@
+
+		ttl := getTTLFromAnnotations(transportServer.Annotations, resource)
+
+		targets := getTargetsFromTargetAnnotation(transportServer.Annotations)
+		if len(targets) == 0 && transportServer.Spec.VirtualServerAddress != "" {
+			targets = append(targets, transportServer.Spec.VirtualServerAddress)
+		}
+		if len(targets) == 0 && transportServer.Status.VSAddress != "" {
+			targets = append(targets, transportServer.Status.VSAddress)
+		}
```
Wdyt of detailing this behavior in the tutorial ?

visokoo replied:
LMK if the examples I added are enough.

mloiseleur commented:
The code lgtm.
Would you please provide a (simple) example of this CR with external dns annotation in the tutorial ? 
This can help newcomers. It has already been done in Traefik or Istio tutorial.

mloiseleur commented:
That's far better, thank you :+1: 
/lgtm
/assign @Raffo

visokoo commented:
@mloiseleur, I synced with master and looks like the `lgtm` to me label got removed. Would you mind readding the tag?

mloiseleur commented:
/lgtm

visokoo commented:
@Raffo is there anything else that we need for this to be merged?

ivankatliarchuk commented:
/lgtm

cc @mloiseleur

mloiseleur commented:
@visokoo you just have to `make generate-flags-documentation` and it should pass. You may have to rebase on the latest master, too.

visokoo commented:
@mloiseleur @ivankatliarchuk, rebased and ran `make generate-flags-documentation`.

Also added @mikejoh's changes from #4996 to this PR as well since that's been merged in.

ivankatliarchuk commented:
Should not force-push. If changes in master, they should not appear in your pull request. Have you rebased correctly?

ivankatliarchuk commented:
Basically, you need to sync your fork with the master and rebase/merge this branch with master

visokoo commented:
> Should not force-push. If changes in master, they should not appear in your pull request. Have you rebased correctly?

@ivankatliarchuk, the force push was for fixing my commit msg that originally contained the `@` symbol. @mikejoh's change was for the `f5-virtualserver` source. I'm mirroring that change to this `f5-transportserver` source.

ivankatliarchuk reviewed (approved):
/lgtm
cc @mloiseleur
